diff --git a/swig/csharp/CMakeLists.txt b/swig/csharp/CMakeLists.txt
index ea9c5bb985..856bfa2ee4 100644
--- a/swig/csharp/CMakeLists.txt
+++ b/swig/csharp/CMakeLists.txt
@@ -1,4 +1,7 @@
 option(GDAL_CSHARP_ONLY "Compile C# bindings on existing GDAL installation" OFF)
+option(GDAL_CSHARP_APPS "Build C# apps" OFF)
+option(GDAL_CSHARP_TESTS "Build C# tests" OFF)
+option(GDAL_CSHARP_BUILD_NUPKG "Build C# tests" OFF)
 
 if (GDAL_CSHARP_ONLY)
   find_package(GDAL REQUIRED)
@@ -18,10 +21,10 @@ set(GDAL_CSHARP_INSTALL_DIR
 
 if (DOTNET_FOUND)
   set(CSHARP_LIBRARY_VERSION
-      "net6.0"
+      "net10.0"
       CACHE STRING ".NET version to be used for libraries")
   set(CSHARP_APPLICATION_VERSION
-      "net6.0"
+      "net10.0"
       CACHE STRING ".NET version to be used for the sample Applications")
 else ()
   set(CSHARP_LIBRARY_VERSION
@@ -278,8 +281,10 @@ function (gdal_csharp_dll)
   endif ()
 
   set(_CSHARP_IMPORTS ${CMAKE_CURRENT_BINARY_DIR}/${_CSHARP_TARGET_SUBDIR}/${_CSHARP_TARGET})
-  if (DOTNET_FOUND)
-    list(APPEND ${CMAKE_CURRENT_BINARY_DIR}/${_PACKAGE_NAME}.${_VERSION_STRING}.nupkg)
+  if(GDAL_CSHARP_BUILD_NUPKG)
+    if (DOTNET_FOUND)
+      list(APPEND ${CMAKE_CURRENT_BINARY_DIR}/${_PACKAGE_NAME}.${_VERSION_STRING}.nupkg)
+    endif ()
   endif ()
 
   install(
@@ -373,7 +378,7 @@ set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES "gdal;ogr;osr;co
 
 # ######################################################################################################################
 # sample commands
-
+if (GDAL_CSHARP_APPS)
 function (gdal_build_csharp_sample)
 
   # setup arguments
@@ -501,7 +506,7 @@ function (gdal_build_csharp_sample)
     DESTINATION ${GDAL_CSHARP_INSTALL_DIR})
 
 endfunction ()
-
+endif()
 # Custom Target to make the C# wrap Libraries and use SWIG to build the .cs files
 add_custom_target(csharp_interface DEPENDS gdalconst_wrap osr_wrap ogr_wrap gdal_wrap)
 
@@ -509,7 +514,7 @@ add_custom_target(csharp_interface DEPENDS gdalconst_wrap osr_wrap ogr_wrap gdal
 add_custom_target(csharp_binding ALL DEPENDS csharp_interface OSGeo.GDAL.CONST OSGeo.GDAL OSGeo.OGR OSGeo.OSR)
 
 if (BUILD_TESTING)
-
+  if (GDAL_CSHARP_APPS)
     # Build the samples
     gdal_build_csharp_sample(
       OUTPUT
@@ -771,6 +776,7 @@ if (BUILD_TESTING)
               OSGeo.GDAL.Samples.GDALWarp
               OSGeo.GDAL.Samples.OGRFeatureEdit
               OSGeo.GDAL.Samples.GetCRSInfo.exe)
+    endif()
 
     # set up the tests
 
